### PR TITLE
Improve VueJS chapter

### PR DESCRIPTION
**Pull request by jamessspanggg**
```
The current VueJS chapter has inconsistencies such as unmatched
curly braces, inconsistent indentation etc. The chapter is also 
missing some important features such as Vue lifecycle hooks.

Let's fix those inconsistencies and add the Vue lifecycle hooks 
description.
```

### CODE REVIEW

File: contents/javascript/Javascript-framework-VueJs.md [hunk 1]
```diff
@@ -124,15 +124,17 @@ In this case, only the `root` component can be accessed in Vue while the rest ar
 
     This follows the [1-way data flow](https://vuejs.org/v2/guide/components-props.html#One-Way-Data-Flow) encouraged by Vue, which
     ensures that data can only be changed by the component itself and also allows bugs to be easily traced in the code.
+    
+    To pass props to a Vue component, `v-bind:prop-name` is used. A demonstration of passing props is shown in the code segment below:
```
Comment (chel-seyy) [inline contents/javascript/Javascript-framework-VueJs.md]: `v-bind:<prop-name>` is clearer.

File: contents/javascript/Javascript-framework-VueJs.md [hunk 2]
```diff
@@ -166,13 +168,17 @@ In this case, only the `root` component can be accessed in Vue while the rest ar
...
     /* Inside todo-list component */
     template: {
-        v-on:increased-count="updateCount"
+      v-on:increased-count="updateCount"
     }
     ```
+   <box type="tip">
+   
+   When the increment item count button is clicked, the `item` component will emit a custom event named `increased-count` while the `todo-list` component listens to the `increased-count` event and executes it's own `updateCount` method.
+   </box>
```
Comment (chel-seyy) [inline contents/javascript/Javascript-framework-VueJs.md:195]: When the ~increment item count~ button is clicked, the `item` component will emit a custom event named `increased-count` while the `todo-list` component listens ~to the `increased-count`~ **for this** event and executes **its** own `updateCount` method.

File: contents/javascript/Javascript-framework-VueJs.md [hunk 3]
```diff
@@ -228,6 +234,21 @@ In this case, only the `root` component can be accessed in Vue while the rest ar
 
 <br>
 
+7. **Lifecycle Hooks**<br>
+    Every Vue instance goes through a series of initialization steps when it is created, i.e. setting up data observation, compiling the template, mounting the instance to the DOM, and updating the DOM when data modifies. Along these steps, Vue runs functions called lifecycle hooks, allowing users to add code at specific steps.<br/><br/>
```
Comment (chel-seyy) [inline contents/javascript/Javascript-framework-VueJs.md]: Along these steps, Vue runs functions **in the background** called lifecycle hooks, allowing users to add code ~at specific steps~ **at each stage that could improve its rendering speed.**

File: contents/javascript/Javascript-framework-VueJs.md [hunk 4]
```diff
@@ -228,6 +234,21 @@ In this case, only the `root` component can be accessed in Vue while the rest ar
...
+    For example, the `created` hook can be used to run code after the Vue instance is created:
```
Comment (chel-seyy) [inline contents/javascript/Javascript-framework-VueJs.md]: hmm.. isn't there `beforeMount` hook for this functionality?

File: contents/javascript/Javascript-framework-VueJs.md [hunk 5]
```diff
@@ -228,6 +234,21 @@ In this case, only the `root` component can be accessed in Vue while the rest ar
...
+    ```js
+    Vue.component('todo-list', {
+      ...
+      created: function() {
+        console.log("to-do list is created.");
+      }
+    })
+    ```
+   There are also other hooks that will be called at different stages of the instance's lifecycle, such as `beforeCreate`, `beforeMount`, `mounted`, `beforeUpdate`, `updated`, `beforeDestroy` and `destroyed`. The lifecycle diagram below shows specific stages when each hook is executed:<br/>
```
Comment (chel-seyy) [inline contents/javascript/Javascript-framework-VueJs.md]: Perhaps can restructure your explanation to support the diagram below?

**Review by chel-seyy — COMMENTED** [review]
Nice work.

File: contents/javascript/Javascript-framework-VueJs.md (hunk 5 above)
Comment (jamessspanggg) [reply contents/javascript/Javascript-framework-VueJs.md]: Thanks for the comments! I have updated the chapter based on your comments. 

I also restructured the explanation for the lifecycle hooks section.

File: contents/javascript/Javascript-framework-VueJs.md [hunk 6]
```diff
@@ -235,8 +235,13 @@ In this case, only the `root` component can be accessed in Vue while the rest ar
...
-    For example, the `created` hook can be used to run code after the Vue instance is created:
+    Every Vue instance goes through a series of initialization steps when it is created, i.e. setting up data observation, compiling the template, mounting the instance to the DOM, and updating the DOM when data modifies. Along these steps, Vue runs functions in the background called lifecycle hooks, allowing users to add code at each stage that could improve its rendering speed.<br/><br/>
+    The following diagram shows all hooks and their specific execution stages:<br/>
+    ![Vue's Lifecycle Diagram](javascript-framework-vue-images/vue-lifecycle-diagram.png "Vue's Lifecycle Diagram")
+    _Figure 1. Vue's Lifecycle Diagram_
+   
+    To run code to a specific stage, you can just define the corresponding hook function and add your code within the function.
```
Comment (chel-seyy) [inline contents/javascript/Javascript-framework-VueJs.md]: run code ~to~ at

**Review by chel-seyy — APPROVED** [review]

File: contents/javascript/Javascript-framework-VueJs.md (hunk 6 above)
Comment (jamessspanggg) [reply contents/javascript/Javascript-framework-VueJs.md]: Thanks for the review!

File: contents/javascript/Javascript-framework-VueJs.md [hunk 7]
```diff
@@ -80,11 +80,11 @@ In this case, only the `root` component can be accessed in Vue while the rest ar
 <b>Step 4:</b> Pass the message to the HTML file using double curly brackets.
 ```HTML
   <div id="root">
-    <h2>{\{message}\{</h2>
+    <h2>{\{message}\}</h2>
```
Comment (ang-zeyu) [inline contents/javascript/Javascript-framework-VueJs.md:83]: A little troublesome, but you could workaround all these occurences temporarily by copying the html output and replacing `{\{` with `{<span>{</span>`. Might be something for another PR if there's too much though

image

File: contents/javascript/Javascript-framework-VueJs.md [hunk 8]
```diff
@@ -97,7 +97,7 @@ In this case, only the `root` component can be accessed in Vue while the rest ar
     ```js
     this.message = 'Hello Space';
     ```
-    When `message` is changed, the view will be re-rendered to show the new message. So you can say, DOM is "reacting" to `message`.
+    When `message` is changed, the view will be re-rendered to show the new message. So you can say, DOM is "reacting" to the changes in `message`.
```
Comment (ang-zeyu) [inline contents/javascript/Javascript-framework-VueJs.md]: Could fix `...the DOM...` here as well

File: contents/javascript/Javascript-framework-VueJs.md [hunk 9]
```diff
@@ -166,13 +168,17 @@ In this case, only the `root` component can be accessed in Vue while the rest ar
       template: {
         <button v-on:click="$emit('increased-count', count+1)">Increment item count</button>
       }
-    }
+    })
 
     /* Inside todo-list component */
     template: {
-        v-on:increased-count="updateCount"
+      v-on:increased-count="updateCount"
```
Comment (ang-zeyu) [inline contents/javascript/Javascript-framework-VueJs.md]: is this an old syntax? The closest I could find in the docs was using `on:` in render functions.
Could fix this as well if so

File: contents/javascript/Javascript-framework-VueJs.md [hunk 10]
```diff
@@ -228,6 +234,24 @@ In this case, only the `root` component can be accessed in Vue while the rest ar
 
 <br>
 
+7. **Lifecycle Hooks**<br>
+    Every Vue instance goes through a series of initialization steps when it is created, i.e. setting up data observation, compiling the template, mounting the instance to the DOM, and updating the DOM when data modifies. Along these steps, Vue runs functions in the background called lifecycle hooks, allowing users to add code at each stage that could improve its rendering speed.<br/><br/>
+    The following diagram shows all lifecycle hooks and their specific execution stages:<br/>
+    ![Vue's Lifecycle Diagram](javascript-framework-vue-images/vue-lifecycle-diagram.png "Vue's Lifecycle Diagram")
```
Comment (ang-zeyu) [inline contents/javascript/Javascript-framework-VueJs.md:250]: The image is a little blurry here. Could resave it at a higher resolution perhaps

**Review by ang-zeyu — COMMENTED** [review]
Looks good 👍

File: contents/javascript/Javascript-framework-VueJs.md (hunk 7 above)
Comment (ang-zeyu) [reply contents/javascript/Javascript-framework-VueJs.md:83]: On second thought, there's an ongoing fix in https://github.com/MarkBind/markbind/issues/762, might be better to do it in a future PR if it needs to be changed again

File: contents/javascript/Javascript-framework-VueJs.md [hunk 11]
```diff
@@ -172,7 +172,7 @@ In this case, only the `root` component can be accessed in Vue while the rest ar
 
     /* Inside todo-list component */
     template: {
-      v-on:increased-count="updateCount"
+      on:increased-count="updateCount"
```
Comment (ang-zeyu) [inline contents/javascript/Javascript-framework-VueJs.md]: Hmm my bad, i meant the `data` object for render functions https://vuejs.org/v2/guide/render-function.html#v-model

Perhaps you could just change it to `template: '...<item v-on:increasedCount="updateCount" v-for="item in items"/>...'`

File: contents/javascript/Javascript-framework-VueJs.md [hunk 12]
```diff
@@ -172,7 +172,7 @@ In this case, only the `root` component can be accessed in Vue while the rest ar
 
     /* Inside todo-list component */
     template: {
-      on:increased-count="updateCount"
+      <item v-on:increased-count="updateCount" v-for="item in items"/>
```
Comment (ang-zeyu) [inline contents/javascript/Javascript-framework-VueJs.md]: `template: '...' `, just noticed the one on top uses `{ ... }` as well, could change that too

Comment (jamessspanggg) [reply contents/javascript/Javascript-framework-VueJs.md]: I standardise to use backticks instead of single quotes, since the top one has multiple lines

File: contents/javascript/Javascript-framework-VueJs.md [hunk 13]
```diff
@@ -124,15 +124,18 @@ In this case, only the `root` component can be accessed in Vue while the rest ar
 
     This follows the [1-way data flow](https://vuejs.org/v2/guide/components-props.html#One-Way-Data-Flow) encouraged by Vue, which
     ensures that data can only be changed by the component itself and also allows bugs to be easily traced in the code.
+    
+    To pass props to a Vue component, `v-bind:<prop-name>` is used. A demonstration of passing props is shown in the code segment below:
```
Comment (openorclose) [inline contents/javascript/Javascript-framework-VueJs.md:144]: Suggestion:

Find another example to show `v-bind` since `v-for` is not introduced anywhere. Alternatively, add a section talking about conditional rendering with `v-if` or `v-for`.

File: contents/javascript/Javascript-framework-VueJs.md [hunk 14]
```diff
@@ -80,11 +80,11 @@ In this case, only the `root` component can be accessed in Vue while the rest ar
...
 ```HTML
   <div id="root">
-    <h2>{\{message}\{</h2>
+    <h2>{\{message}\}</h2>
   </div>
 ```
 
-<b>Step 5:</b> Open the brower and we will see "Hello World" being displayed:
+<b>Step 5:</b> Open the browser and we will see "Hello World" being displayed:
 > <h2>Hello World</h2>
```
Comment (openorclose) [inline contents/javascript/Javascript-framework-VueJs.md:88]: Maybe replace this with an actual screenshot?

File: contents/javascript/Javascript-framework-VueJs.md [hunk 15]
```diff
@@ -207,13 +210,13 @@ In this case, only the `root` component can be accessed in Vue while the rest ar
 
     ```js
     watch: {
-        totalCount: function() {
-            let result = 0
-            this.items.forEach((item) => result += item.count);
-            this.totalCount = result;
+      totalCount: function() {
+        let result = 0
+        this.items.forEach((item) => result += item.count);
```
Comment (openorclose) [inline contents/javascript/Javascript-framework-VueJs.md]:
```suggestion
        this.items.forEach((item) => result += item.count);
```

Not Vue related, but can be more idiomatic using `Array.prototype.reduce` to sum instead,

**Review by openorclose — CHANGES_REQUESTED** [review]
Just some small issues, otherwise LGTM!

File: contents/javascript/Javascript-framework-VueJs.md (hunk 14 above)
Comment (jamessspanggg) [reply contents/javascript/Javascript-framework-VueJs.md:88]: don't think this is necessary tho, the point to bring across is that `Hello World` will be displayed

**Review by eugenepeh — COMMENTED** [review]
* The div is not closed 
image

* Perhaps standardize all indentation to 2 spaces?
image

**Review by eugenepeh — APPROVED** [review]

File: contents/javascript/Javascript-framework-VueJs.md (hunk 12 above)
Comment (ang-zeyu) [reply contents/javascript/Javascript-framework-VueJs.md]: LGTM 👍

**Review by ang-zeyu — COMMENTED** [review]
Just noticed this section was misformatted as well:

image

Likely cause would be the extra indent on the code blocks in that section

**Comment by jamessspanggg** [comment]
> Likely cause would be the extra indent on the code blocks in that section

Good catch. Fixed the indentation issue.

**Review by ang-zeyu — APPROVED** [review]
Looks good 👍 
Just one final nit

Could rewrite this warning here that `{\{` should be `{{` and  `}\}` should be `}}` due to the limitations of MarkBind currently, as the reader may misinterpret that `{\{` should be `<nothing>`.

Could use `{<span>{</span>` for it

image

**Comment by jamessspanggg** [comment]
> Could rewrite this warning here that `{\{` should be `{{` and `}\}` should be `}}` due to the limitations of MarkBind currently, as the reader may misinterpret that `{\{` should be `<nothing>`.
> 
> Could use `{<span>{</span>` for it

In this case, might as well use `{<span>{</span>` for all?

**Comment by jamessspanggg** [comment]
> Could use `{<span>{</span>` for it

I've tried and apparently it doesn't work tho:
Screen Shot 2020-02-19 at 5 01 41 PM

**Comment by ang-zeyu** [comment]
> I've tried and apparently it doesn't work tho:
> Screen Shot 2020-02-19 at 5 01 41 PM

My bad, try this `<code>{<span>{</span></code>`, forgot to mention replacing backticks with `<code>`

**Review by iskandarzulkarnaien — APPROVED** [review]
LGTM

**Comment by jamessspanggg** [comment]
> My bad, try this `<code>{<span>{</span></code>`, forgot to mention replacing backticks with `<code>`

Ah works now, thanks so much!
